etl/churn.kp/knowledge.md

```diff
@@ -433,182 +401,188 @@ def get_newest_per_client(df):
     )
     return selectable_by_client.filter(selectable_by_client['row_number'] == 1)
 
-DO_UPLOAD = True
-
-world_start = '20151101'
-today = _datetime.utcnow()
-todays = fmt(today)
-wsd = _datetime.strptime(world_start, "%Y%m%d")
-week_start_date = wsd
-week_end_date = week_start_date + timedelta(6)
-week_start = fmt(week_start_date)
-week_end = fmt(week_end_date)
 
-client = boto3.client('s3', 'us-west-2')
-transfer = S3Transfer(client)
-bucket = "net-mozaws-prod-us-west-2-pipeline-analysis"
-
-MAX_SUBSESSION_LENGTH = 60 * 60 * 48 # 48 hours in seconds.
+def upload_to_s3(records, churn_outfile):
+    client = boto3.client('s3', 'us-west-2')
+    transfer = S3Transfer(client)
+    churn_bucket = "net-mozaws-prod-us-west-2-pipeline-analysis"
+    churn_filepath = "amiyaguchi/churn"
+
+    print "{}: Writing output to {}".format(_datetime.utcnow(), churn_outfile)
+
+    # Write the file out as gzipped csv
+    with gzip.open(churn_outfile, 'wb') as fout:
+        fout.write(",".join(record_columns) + "\n")
+        print "{}: Wrote header to {}".format(_datetime.utcnow(), churn_outfile)
+        for r in records:
+            try:
+                fout.write(csv(r))
+                fout.write("\n")
+            except UnicodeEncodeError as e:
+                print("{}: Error writing line: {} // {}"
+                      .format(_datetime.utcnow(), e, r))
+        print "{}: finished writing lines".format(_datetime.utcnow())
+
+    # Now upload it to S3:
```

Comment:
Uploads to s3 can cause the job to fail if the object already exists. I should check for existence in both instances of uploading.